### PR TITLE
NEW Work with new HTTPCacheControl class (2.x)

### PR DESCRIPTION
**Pull request by chillu**
Squashed and modified version of https://github.com/silverstripe/silverstripe-controllerpolicy/pull/21

### CODE REVIEW

**Comment by chillu** [comment]
Tried to run the same tests as on #21, but I'm always getting a session cookie back. At first I blamed the CWP recipe and hybridsessions, but I can also reproduce this on a stock 4.x install: https://github.com/silverstripe/silverstripe-framework/issues/8267

File: tests/CachingPolicyTest.php [hunk 1]
```diff
@@ -82,16 +88,20 @@ public function testConfigured()
...
-            $response->getHeader('Vary'),
-            'Header appears as configured'
+        $directives = explode(',', $response->getHeader('Cache-Control'));
+        $directives = array_map('trim', $directives);
+
+        $this->assertCount(2, $directives);
+        $this->assertContains('max-age=999', $directives);
+        $this->assertContains('must-revalidate', $directives);
+
+        $vary = $this->getVaryAsArray($response->getHeader('Vary'));
```
[the remark this answers is not in the document]
Comment (chillu) [reply tests/CachingPolicyTest.php]: Good point, victim of only changing part of the test - changed now

File: tests/CachingPolicyTest.php [hunk 2]
```diff
@@ -82,16 +88,20 @@ public function testConfigured()
...
+        $this->assertArraySubset(
+            array_keys(HTTPCacheControlMiddleware::config()->get('defaultVary')),
+            $vary,
+            'Retains default Vary'
```
[the remark this answers is not in the document]
Comment (chillu) [reply tests/CachingPolicyTest.php]: That's not how I interpret `$message` wording, but I can't find any guidance either way in https://phpunit.readthedocs.io/en/7.1/assertions.html. It just says "Reports an error identified by $message ..."

File: tests/CachingPolicyTest.php [hunk 3]
```diff
@@ -82,16 +88,20 @@ public function testConfigured()
...
         );
+        $this->assertContains('X-EyeColour', $vary, 'Adds custom vary');
```
[the remark this answers is not in the document]
Comment (chillu) [reply tests/CachingPolicyTest.php]: Taken over from existing test, those should have better names anyway, but I'm not going to go out of my way on this deprecated module ;)

File: tests/CachingPolicyTest.php [hunk 4]
```diff
@@ -127,26 +138,17 @@ public function testUnrelated()
...
+        $vary = $this->getVaryAsArray($response->getHeader('Vary'));
+        $this->assertEquals($vary, $defaultVary, 'Headers on unrelated controller are unaffected');
     }
 
-    public function testModificationDateFromDataObjects()
+    /**
+     * @param string $varyStr
+     * @return array
+     */
+    protected function getVaryAsArray($varyStr)
```
[the remark this answers is not in the document]
Comment (chillu) [reply tests/CachingPolicyTest.php]: Doneburgers

File: README.md [hunk 5]
```diff
@@ -7,6 +7,12 @@
 
 ## Summary
 
+**This module is deprecated for usage with SilverStripe 3.7+**.
+The [HTTPCacheControl API](https://docs.silverstripe.org/en/3/developer_guides/performance/http_cache_headers/)
```
[the remark this answers is not in the document]
Comment (chillu) [reply README.md]: Done

File: src/Policies/CachingPolicy.php [hunk 6]
```diff
@@ -98,133 +99,59 @@ public function getVary()
      */
     public function applyToResponse($originator, HTTPRequest $request, HTTPResponse $response)
     {
-        $cacheAge = $this->getCacheAge();
-        $vary = $this->getVary();
+        $cacheAge = $this->cacheAge;
+        $vary = $this->vary;
 
-        // Allow overriding max-age from the object hooked up to the policed controller.
+                // Allow overriding max-age from the object hooked up to the policed controller.
```
[the remark this answers is not in the document]
Comment (chillu) [reply src/Policies/CachingPolicy.php]: Done

File: tests/CachingPolicyTest.php [hunk 7]
```diff
@@ -39,6 +39,12 @@ protected function setUp()
         parent::setUp();
 
         Config::modify()->set(CachingPolicy::class, 'disable_cache_age_in_dev', false);
+
+        // Set to disabled at null forcing level, overrides dev mode defaults
+        HTTPCacheControlMiddleware::config()
+            ->set('defaultState', 'disabled')
```
[the remark this answers is not in the document]
Comment (chillu) [reply tests/CachingPolicyTest.php]: I've removed that whole setting, it's disabled in both dev mode and the class default. The `defaultForcingLevel` is the important bit

**Comment by chillu** [comment]
@dhensby Responded to all of your peer review feedback

File: tests/CachingPolicyTest.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (dhensby) [reply tests/CachingPolicyTest.php]: Yeah, so "testX failed to assert ArraySubset. Retains default vary" is basically whe the failure would say (you don't get a message on success).

**Comment by chillu** [comment]
Since Dan approved the PR, and I've since fixed the tests, I'm merging.

**Pull request merged**